Userland/Libraries/LibGL/SoftwareRasterizer.cpp

```diff
@@ -0,0 +1,257 @@
+{
+    return ((c.x - a.x) * (b.y - a.y) - (c.y - a.y) * (b.x - a.x)) / 2;
+}
+
+constexpr static Gfx::RGBA32 to_rgba32(f32x4 v)
+{
+    u8 r = clamp(v[0], 0.0f, 1.0f) * 255;
+    u8 g = clamp(v[1], 0.0f, 1.0f) * 255;
+    u8 b = clamp(v[2], 0.0f, 1.0f) * 255;
+    u8 a = clamp(v[3], 0.0f, 1.0f) * 255;
+    return a << 24 | b << 16 | g << 8 | r;
+}
+
+template<typename PS>
+static void rasterize_triangle(Gfx::Bitmap& render_target, const GLTriangle& triangle, PS pixel_shader)
+{
+    // Since the algorithm is based on blocks of uniform size, we need
+    // to ensure that our render_target size is actually a multiple of the block size
+    VERIFY((render_target.width() % RASTERIZER_BLOCK_SIZE) == 0);
+    VERIFY((render_target.height() % RASTERIZER_BLOCK_SIZE) == 0);
+
+    // Calculate area of the triangle for later tests
+    FloatVector2 v0 = { triangle.vertices[0].x, triangle.vertices[0].y };
+    FloatVector2 v1 = { triangle.vertices[1].x, triangle.vertices[1].y };
+    FloatVector2 v2 = { triangle.vertices[2].x, triangle.vertices[2].y };
+
+    float area = triangle_area(v0, v1, v2);
+    if (area == 0)
+        return;
+
+    float one_over_area = 1 / area;
+
+    // Obey top-left rule:
+    // This sets up "zero" for later pixel coverage tests.
+    // Depending on where on the triangle the edge is located
+    // it is either tested against 0 or float epsilon, effectively
+    // turning "< 0" into "<= 0"
+    float constexpr epsilon = AK::NumericLimits<float>::epsilon();
+    f32x4 zero { epsilon, epsilon, epsilon, 0.0f };
+    if (v1.y > v0.y || (v1.y == v0.y && v1.x < v0.x))
+        zero[2] = 0;
+    if (v2.y > v1.y || (v2.y == v1.y && v2.x < v1.x))
+        zero[0] = 0;
+    if (v0.y > v2.y || (v0.y == v2.y && v0.x < v2.x))
+        zero[1] = 0;
+
+    // This function calculates the barycentric coordinates for the pixel relative to the triangle.
+    auto barycentric_coordinates = [v0, v1, v2, one_over_area](float x, float y) -> f32x4 {
+        FloatVector2 p { x, y };
+        return f32x4 {
+            triangle_area(v1, v2, p) * one_over_area,
+            triangle_area(v2, v0, p) * one_over_area,
+            triangle_area(v0, v1, p) * one_over_area,
+            0.0f
+        };
+    };
+
+    // This function tests whether a point lies within the triangle
+    auto test_point = [zero](f32x4 point) -> bool {
+        return point[0] >= zero[0]
+            && point[1] >= zero[1]
+            && point[2] >= zero[2];
+    };
+
+    // Calculate bounds
+    FloatVector2 min { v0 };
+    FloatVector2 max { v0 };
+    if (v1.x > max.x)
+        max.x = v1.x;
+    if (v2.x > max.x)
+        max.x = v2.x;
+    if (v1.y > max.y)
+        max.y = v1.y;
+    if (v2.y > max.y)
+        max.y = v2.y;
+    if (v1.x < min.x)
+        min.x = v1.x;
+    if (v2.x < min.x)
+        min.x = v2.x;
+    if (v1.y < min.y)
+        min.y = v1.y;
+    if (v2.y < min.y)
+        min.y = v2.y;
```

Comment:
Could you please change this to be a chain of something like `max.x = AK::max(max.x, AK::max(v1.x, v2.x));` ? GCC compiles this much better than the chain of conditional assignments (https://godbolt.org/z/xqWYdd3xs shows that the std::max version generates cmovs while the if statement chain generates lots of jumps). I think gcc can't figure out that these can be done branchlessly because you don't have else clauses for these if statements.

Reply:
I've found that gcc has issues converting conditional assignments into branchless assignments unless you have both if and else clauses assigning to the same variable.